src/jit/optimizer.cpp

```diff
@@ -6789,6 +6789,12 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                 return false;
             }
 
+            // Don't hoist constants or nodes that have been assigned a constant value.
+            if (m_compiler->vnStore->IsVNConstant(node->gtVNPair.GetLiberal()))
```

Comment:
At a minimum this needs a much better comment, as it is unclear *why* it should not hoist constants.
But I wonder whether this is excessively conservative. Have you run diffs? @mikedn said here: https://github.com/dotnet/coreclr/issues/26417#issuecomment-525884714 that perhaps it should be checking the Conservative VN for this case.

Reply:
Nothing good happens by placing hoisting a constant into the loop pre-header here.

The CSE logic currently has code to avoid performing  CSE's on constants.
Unfortunately it checks the Conservative VN, which allows this bug to creep in:

optcse.cpp line 649
```
                // Don't CSE constant values, instead let the Value Number
                // based Assertion Prop phase handle them.  Here, unlike
                // the rest of optCSE, we use the conservative value number
                // rather than the liberal one, since the conservative one
                // is what the Value Number based Assertion Prop will use
                // and the point is to avoid optimizing cases that it will
                // handle.
                //
                if (vnStore->IsVNConstant(vnStore->VNConservativeNormalValue(tree->gtVNPair)))
                {
                    continue;
                }
```

Reply:
We will still hoist invariant expression that contain constant values, just not no a solitary constant, or in the case of this bug, a GT_CLS_VAR that has been assigned a constant value.
For hoisting we are simply looking at the Liberal Value number has hoisting it  if it identifies as a constant.

Reply:
> We will still hoist invariant expressions that contain constant values, 

This is no longer the case with the updated fix, right?

Reply:
Ah, I see that there was another updated that deals with this.